src/stream/src/executor/backfill/arrangement_backfill.rs

```diff
@@ -0,0 +1,530 @@
+            "Backfill has already finished and forward messages directly to the downstream"
+        );
+
+        // Wait for first barrier to come after backfill is finished.
+        // So we can update our progress + persist the status.
+        while let Some(Ok(msg)) = upstream.next().await {
+            if let Some(msg) = mapping_message(msg, &self.output_indices) {
+                // If not finished then we need to update state, otherwise no need.
+                if let Message::Barrier(barrier) = &msg && !is_completely_finished {
+                    // If snapshot was empty, we do not need to backfill,
+                    // but we still need to persist the finished state.
+                    // We currently persist it on the second barrier here rather than first.
+                    // This is because we can't update state table in first epoch,
+                    // since it expects to have been initialized in previous epoch
+                    // (there's no epoch before the first epoch).
+                    if is_snapshot_empty {
+                        let finished_state = construct_initial_finished_state(pk_in_output_indices.len());
+                        for vnode in upstream_table.vnodes().iter_vnodes() {
+                            backfill_state.update_progress(vnode, BackfillProgressPerVnode::InProgress(finished_state.clone()));
+                        }
+                    }
+
+                    persist_state_per_vnode(
+                        barrier.epoch,
+                        &mut self.state_table,
+                        false,
+                        &mut backfill_state,
+                        &mut committed_progress,
+                        &mut temporary_state,
+                    ).await?;
+
+                    self.progress.finish(barrier.epoch.curr);
+                    yield msg;
+                    break;
+                }
+                yield msg;
+            }
+        }
+
+        // After progress finished + state persisted,
+        // we can forward messages directly to the downstream,
+        // as backfill is finished.
+        #[for_await]
+        for msg in upstream {
+            if let Some(msg) = mapping_message(msg?, &self.output_indices) {
+                if let Message::Barrier(barrier) = &msg {
+                    self.state_table.commit_no_data_expected(barrier.epoch);
+                }
+                yield msg;
+            }
+        }
+    }
+
+    /// Read snapshot per vnode.
+    /// These streams should be sorted in storage layer.
+    /// 1. Get row iterator / vnode.
+    /// 2. Merge it with `select_all`.
+    /// 3. Change it into a chunk iterator with `iter_chunks`.
+    /// This means it should fetch a row from each iterator to form a chunk.
+    ///
+    /// NOTE(kwannoel): We interleave at chunk per vnode level rather than rows.
+    /// This is so that we can compute `current_pos` once per chunk, since they correspond to 1
+    /// vnode.
```

Comment:
Take note of this. Instead of interleaving a row at a time from each vnode, we interleave chunks per vnode.

Vnode needs to be computed to update backfill position for that vnode.